### DISCUSSION TITLE
NPM listening on multiple ports (7777/TCP & 7778/TCP)

### DISCUSSION THREAD

**Discussion by latency0ms** [NginxProxyManager/nginx-proxy-manager#1632] [Q&A]
Dear NPM Community

I am facing a problem that makes my head explode 🤯. I want NPM to listen on two individual ports (7777/TCP & 7778/TCP) and then forward the traffic to the corresponding web servers. I assume that this can be done via the advanced configuration option.

For a better understanding I have created this layout: https://abload.de/img/2021-12-04_1424108tkvo.png

Any help to solve this problem will be highly appreciated.

Thank you.
![2021-12-04_142410](https://user-images.githubusercontent.com/55700671/144711236-dad31e48-b7d7-4dbb-b0e5-6ed4fdf4b36b.png)

**Comment by chaptergy** [comment]
If you want it just on these ports, and no longer on port `80` or `443`, you can change one of the ports to be e.g. `7777` (depending on whether you want ssl encryption) via the docker-compose and then create a stream host for port `7778` which points to `localhost:80` or `localhost:443` depending on which one you changed (since the port mapping is done in docker you need to use the inside port here, not the outside one when using `localhost`). If you want to keep port `80` and `443`, just create two stream hosts to one of the localhost ports.

You can then create normal proxy hosts as you normaly would for your domains, since you have now forwarded that port to one of the default ports.

**Reply by latency0ms** [reply]
Thanks @chaptergy 

I also have setup several other 80 and 443 Proxy Hosts using custom SSL certificates, therefore I am afraid that your solution cannot be applied here?

Do I really need to set up a standalone NGINX reverse proxy for this purpose or is there another way?

**Reply by chaptergy** [reply]
As I said you could also set up two stream hosts to forward both port `7777` and `7778` to one of the default ports `80` or `443` and set them up as normal proxy hosts. However if you want to use the same domain both on the default ports and on your custom ports to point to different things, that is not possible using npm. By that I mean forwarding `app.example.com:443` to 1.2.3.4 and `app.example.com:7777` to 6.7.8.9 is not possible. You would either need to create your own nginx config file or use a separate instance.

**Reply by latency0ms** [reply]
I have the impression that I can't solve this with NPM and need to setup a classic NGINX installation instead.

**Reply by latency0ms** [reply]
@chaptergy I have now setup a new NGINX proxy server, do you maybe have an idea on how to create a working config for my scenario?

I came up with this idea:

```nginx
server {
  listen 7777;

  server_name app1.mydomain.com;

  location / {
      proxy_pass https://192.168.1.10:7777/;
  }
}
server {
  listen 7778;

  server_name app1.mydomain.com;

  location / {
      proxy_pass https://192.168.1.11:7778/;
  }
}
```

**Reply by chaptergy** [reply]
It's only possible if the service you are proxying uses the HTTP protocol or a protocol which has SNI (Server name indication). Otherwise you can't use domains to route the traffic. But as you used "Webserver" in your diagram, I assumed it was HTTP. Otherwise the config you posted is how I would do it. But are you sure the app at `192.168.1.11:7778` and the other location expects ssl encryption (https)?

**Reply by latency0ms** [reply]
These are real webservers (App1, App2, App3).

On the App1 webserver there is a basic http website available, I can access the website over LAN, but the proxy server is returning the internal IP address (192.168.1.10) in the addressbar of the browser. 

![image](https://user-images.githubusercontent.com/55700671/144727459-7de297b5-6317-4f10-916b-a618ca5b0261.png)


This is the config I use:

```nginx
server {
  listen 80;

  server_name app1.mydomain.com;

  location / {
      proxy_pass http://192.168.1.10:80/;
  }
}
```

**Reply by latency0ms** [reply]
Ok, now I can access the basic website over http by adding this line to my config `proxy_set_header Host $host;`

![image](https://user-images.githubusercontent.com/55700671/144728216-c51a831c-c7d8-45bd-a803-dc12a0c785f4.png)

**Reply by latency0ms** [reply]
Now I have to figure out on how to get 7778 (https) working. Over LAN I can access https://192.168.1.10:7778/ but my config is not forwarding for some reason:

```nginx
server {
  listen 7777;

  server_name app1.mydomain.com;

  location / {
      proxy_pass http://192.168.1.10:7777/;
      proxy_set_header Host $host;
  }
}
server {
  listen 7778;
  listen 443;

  server_name app1.mydomain.com;

  location / {
      proxy_pass https://192.168.1.10:7778/;
      proxy_set_header Host $host;
  }
}
```
A valid ssl certificate is installed on the App1 server.

**Reply by latency0ms** [reply]
Got this thing flying with the new Nginx setup. If someone is interested in the config, let me know.
Thanks

**Reply by LForLuketastic** [reply]
@latency0ms I'm really interested in the config! I would welcome this very much

**Reply by latency0ms** [reply]
```nginx
###################################### LAB.1 Development ######################################
#----- OLAP CONFIG
server {
	listen 7778 ssl;
	server_name app1-olap.mydomain.com;
	ssl_certificate /etc/nginx/sslcerts/cert_jdx.pem;
	ssl_certificate_key /etc/nginx/sslcerts/cert_jdx.pem;

location / {
	proxy_pass https://192.168.1.10:7778/;
	proxy_set_header Host $host;
	
    }
}
#----- SSL CONFIG
server {
	listen 443 ssl;
	
	server_name                     app1.mydomain.com;
	ssl_certificate                 /etc/nginx/sslcerts/cert_jdx.pem;
	ssl_certificate_key             /etc/nginx/sslcerts/cert_jdx.pem;
	ssl_session_cache               builtin:1000 shared:SSL:10m;
	ssl_protocols                   TLSv1 TLSv1.1 TLSv1.2;
	ssl_ciphers                     HIGH:!aNULL:!eNULL:!EXPORT:!CAMELLIA:!DES:!MD5:!PSK:!RC4;
	ssl_prefer_server_ciphers       on;

location / {

	# Set the proxy headers
	proxy_set_header                Host $host;
	proxy_set_header                X-Real-IP $remote_addr;
	proxy_set_header                X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header                X-Forwarded-Proto $scheme;

	# Configure which address the request is proxied to
	proxy_pass                      http://192.168.1.10:80;
	proxy_read_timeout              90;

	# Security headers
	add_header Strict-Transport-Security "max-age=31536000; includeSubDomains; preload";
	add_header X-Frame-Options DENY;
	add_header X-Content-Type-Options nosniff;
	add_header X-XSS-Protection "1; mode=block";
	add_header Referrer-Policy "origin";

	# Add the trailing slash
	rewrite ^([^.]*[^/])$ $1/ permanent;
	
	}
}
############################################################################################
```

**Reply by LForLuketastic** [reply]
@latency0ms Thank you so much! Can i contact you privatly? i got a question, that making my head explode about nginx

**Reply by DaKa22** [reply]
@latency0ms pero no se pudo configurar desde el NPM ? , estoy pasando por la misma situacion y no quiero recurrir a un NGINX tradicional y dar solucion desde el NPM
